cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/PipelineCondition.java

```diff
@@ -16,17 +16,93 @@
+import co.cask.cdap.internal.io.SchemaTypeAdapter;
+import com.google.common.collect.SetMultimap;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
 
+import java.util.HashMap;
+import java.util.Map;
 import javax.annotation.Nullable;
 
 /**
  * Represents conditions in the data pipelines.
  */
-public class PipelineCondition implements Predicate<WorkflowContext> {
+public class PipelineCondition extends AbstractCondition {
+
+  // This is only visible during the configure time, not at runtime.
+  private final BatchPhaseSpec phaseSpec;
+  private Metrics metrics;
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .registerTypeAdapter(SetMultimap.class, new SetMultimapCodec<>())
+    .create();
+
+  public PipelineCondition(BatchPhaseSpec spec) {
+    this.phaseSpec = spec;
+  }
+
+  @Override
+  protected void configure() {
+    setName(phaseSpec.getPhaseName());
+    setDescription("Condition phase executor. " + phaseSpec.getPhaseName());
+
+    // add source, sink, transform ids to the properties. These are needed at runtime to instantiate the plugins
+    Map<String, String> properties = new HashMap<>();
+    properties.put(Constants.PIPELINEID, GSON.toJson(phaseSpec));
+    setProperties(properties);
+  }
+
   @Override
   public boolean apply(@Nullable WorkflowContext input) {
-    return false;
+    if (input == null) {
+      // should not happen
+      throw new IllegalStateException("WorkflowContext for the Condition cannot be null.");
+    }
+    Map<String, String> properties = input.getConditionSpecification().getProperties();
+    BatchPhaseSpec phaseSpec = GSON.fromJson(properties.get(Constants.PIPELINEID), BatchPhaseSpec.class);
+    PipelinePhase phase = phaseSpec.getPhase();
+    StageSpec stageSpec = phase.iterator().next();
+    PluginContext pluginContext = new PipelinePluginContext(input, metrics, phaseSpec.isStageLoggingEnabled(),
+                                                            phaseSpec.isProcessTimingEnabled());
+
+    MacroEvaluator macroEvaluator = new DefaultMacroEvaluator(input.getToken(), input.getRuntimeArguments(),
+                                                              input.getLogicalStartTime(), input, input.getNamespace());
+
+    try {
+      Condition condition = pluginContext.newPluginInstance(stageSpec.getName(), macroEvaluator);
+      BasicArguments arguments = new BasicArguments(input.getToken(), input.getRuntimeArguments());
+      ConditionContext conditionContext = new BasicConditionContext(input, metrics, stageSpec, arguments);
+      boolean result = condition.apply(conditionContext);
+      WorkflowToken token = input.getToken();
+      if (token == null) {
+        throw new IllegalStateException("WorkflowToken cannot be null when Condition is executed through Workflow.");
+      }
+      for (Map.Entry<String, String> entry : arguments.getAddedArguments().entrySet()) {
```

Comment:
don't think this is needed, as ConditionContext can only get Arguments but not SettableArguments